### PR TITLE
Ethon instrumentation

### LINKED ISSUE #2187: Add Instrumentation for Ethon
Add support for [ethon](https://rubygems.org/gems/ethon)

### PR DESCRIPTION
**Pull request by fallwith**
add instrumentation for Ethon

resolves #2187

### CODE REVIEW

File: newrelic.yml [hunk 1]
```diff
@@ -102,7 +102,7 @@ common: &default_settings
   # Specify a list of constants that should prevent the agent from starting
   # automatically. Separate individual constants with a comma ,. For example,
   # "Rails::Console,UninstrumentedBackgroundJob".
-  # autostart.denylisted_constants: Rails::Console
+  # autostart.denylisted_constants: Rails::Command::ConsoleCommand,Rails::Command::CredentialsCommand,Rails::Command::Db::System::ChangeCommand,Rails::Command::DbConsoleCommand,Rails::Command::DestroyCommand,Rails::Command::DevCommand,Rails::Command::EncryptedCommand,Rails::Command::GenerateCommand,Rails::Command::InitializersCommand,Rails::Command::NotesCommand,Rails::Command::RoutesCommand,Rails::Command::SecretsCommand,Rails::Console,Rails::DBConsole
```
Comment (hannahramadan) [inline newrelic.yml:105]: Since we have automation to update newrelic.yml, we will lose these changes unless updated in `default_source`!

Comment (fallwith) [reply newrelic.yml:105]: Thanks @hannahramadan. The changes made to `newrelic.yml` in this PR were from running the rake task to autogenerate a new file. I likely should have simply withheld the changes from the PR, but it's also not going to hurt anything to have it. I kinda favor `newrelic.yml` be generated on a per PR basis in the future. The `default_source.rb` file has all of the content I need, and I'm okay with the pending `newrelic.yml` regeneration driven by the next release.

Comment (hannahramadan) [reply newrelic.yml:105]: Ooh totally! Agree that running on a PR basis sounds nicer. Having that workflow run on only pushes to `main` leaves room for surprises we might not want. Making a ticket for this.

**Review by kaylareopelle — CHANGES_REQUESTED** [review]

File: lib/new_relic/agent/http_clients/ethon_wrappers.rb [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        def format_key(key)
+          key.tr('-', '_').downcase
+        end
+      end
+
+      class EthonHTTPRequest < AbstractRequest
+        attr_reader :uri
+
+        DEFAULT_ACTION = 'unknownaction'
+        DEFAULT_HOST = 'unknownhost'
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/http_clients/ethon_wrappers.rb]: We discussed this one elsewhere. `Ethon::Easy` doesn't bother with the action in many cases. So let's just use `'GET'` as a default name to better represent what Curl is doing under the hood.

And rename `'unknownhost'` to `'UNKNOWN_HOST'`.

Addressed with [30dc1fe](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/30dc1fea0b3e98b62cfbf763f734aa3d839f2e1f)

File: lib/new_relic/agent/instrumentation/ethon.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+  executes do
+    NewRelic::Agent.logger.info('Installing ethon instrumentation')
+  end
+
+  executes do
+    if use_prepend?
+      prepend_instrument Ethon::Easy, NewRelic::Agent::Instrumentation::Ethon::Easy::Prepend
+      prepend_instrument Ethon::Multi, NewRelic::Agent::Instrumentation::Ethon::Multi::Prepend
+    else
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/ethon.rb]: Thanks for catching this one! Addressed with [30dc1fe](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/30dc1fea0b3e98b62cfbf763f734aa3d839f2e1f)

File: lib/new_relic/agent/instrumentation/ethon.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    defined?(Ethon) && Gem::Version.new(Ethon::VERSION) >= Gem::Version.new('0.12.0')
+  end
+
+  executes do
+    NewRelic::Agent.logger.info('Installing ethon instrumentation')
+  end
+
+  executes do
+    if use_prepend?
+      # NOTE: to prevent a string like 'Ethon::Easy' from being converted into
```
Comment (hannahramadan) [inline lib/new_relic/agent/instrumentation/ethon.rb:22]: nice comment I didn't know `prepend_instrument` supported a 3rd arg

**Review by hannahramadan — DISMISSED** [review]
Thanks for reviewing live ✨

**Review by kaylareopelle — DISMISSED** [review]

File: test/new_relic/http_client_test_cases.rb [hunk 5]
```diff
@@ -531,27 +506,46 @@ def test_still_records_tt_node_when_request_fails
...
-      evil_server.stop
-    end
+  def typhoeus?
+    client_name == 'Typhoeus'
+  end
+
+  def perform_last_node_assertions
+    last_node = find_last_transaction_node()
+
+    expected_name = typhoeus? ? 'Ethon' : client_name
```
Comment (kaylareopelle) [inline test/new_relic/http_client_test_cases.rb:536]: Should this still be there?

Comment (fallwith) [reply test/new_relic/http_client_test_cases.rb]: Whoops! Lots of reversions needed for this file too. Thank you! Addressed with [914302d](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/914302d40142f9f396afa4d839b42c875f463da2)

File: test/new_relic/http_client_test_cases.rb [hunk 6]
```diff
@@ -531,27 +506,42 @@ def test_still_records_tt_node_when_request_fails
...
-      last_node = find_last_transaction_node()
+    perform_last_node_assertions
 
-      assert_equal("External/localhost/#{client_name}/GET", last_node.metric_name)
+    evil_server.stop
+  end
 
-      evil_server.stop
-    end
+  def typhoeus?
```
Comment (kaylareopelle) [inline test/new_relic/http_client_test_cases.rb:529]: Do we still need this method?

**Review by kaylareopelle — DISMISSED** [review]
This is good to merge. I saw two small cleanup things.

File: lib/new_relic/agent/instrumentation/ethon.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      #       'Ethon/Easy', a 3rd argument is supplied to `prepend_instrument`
```
[the remark this answers is not in the document]
Comment (fallwith) [reply lib/new_relic/agent/instrumentation/ethon.rb]: Thank you! Corrected with [285de8a](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/285de8a22b4154a62ecace9ce9856aca7f741161)

File: test/new_relic/http_client_test_cases.rb (hunk 6 above)
Comment (fallwith) [reply test/new_relic/http_client_test_cases.rb]: Good call here. Now that we've reverted the stacked Typhoeus/Ethon instrumentation, this helper method is down to just a single caller. Refactored the method out with [285de8a](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/285de8a22b4154a62ecace9ce9856aca7f741161) to go back to the way `dev` currently has it.

**Comment by fallwith** [comment]
> This is good to merge. I saw two small cleanup things.

Thanks very much! Cleanup addressed with [285de8a](https://github.com/newrelic/newrelic-ruby-agent/pull/2260/commits/285de8a22b4154a62ecace9ce9856aca7f741161)

**Review by kaylareopelle — APPROVED** [review]
🦅

**Pull request merged**